libia2/include/ia2_compartment_init.inc

```diff
@@ -60,3 +65,135 @@ void COMPARTMENT_IDENT(init_tls)(void) {
   };
   dl_iterate_phdr(protect_tls_pages, &args);
 }
+
+/* Magic symbol from the linker that marks the start of the ELF headers */
+extern Elf64_Ehdr __ehdr_start __attribute__((weak));
+
+struct FinalizerInfo {
+  uint64_t fini_offset;
+  uint64_t fini_array_offset;
+  uint64_t fini_array_size;
+};
+
+/* Filled in by runtime initialization with contents of  .fini_array and
+ * .fini */
+static volatile struct FinalizerInfo finalizers
+    __attribute__((__section__(".data.rel.ro.1"))) = {0};
+
+extern void COMPARTMENT_IDENT(__wrap_ia2_compartment_destructor)(void);
+
+__attribute__((no_sanitize("undefined"))) __attribute__((used)) void
+COMPARTMENT_IDENT(ia2_compartment_destructor)(void) {
+  static bool ran_destructors = false;
+  if (!&__ehdr_start || ran_destructors)
+    return;
+
+  void (**fini_array)(void) = (void (**)(void))(((uint8_t *)&__ehdr_start) +
+                                                finalizers.fini_array_offset);
+  for (int i = 0; i < (finalizers.fini_array_size / sizeof(void *)); i++) {
+    fini_array[i]();
+  }
+  if (finalizers.fini_offset) {
+    void (*fini)(void) =
+        (void (*)(void))(((uint8_t *)&__ehdr_start) + finalizers.fini_offset);
+    fini();
+  }
+
+  ran_destructors = true;
+}
+
+/* This pointer MUST be in a read-only (.data.rel.ro) section. */
+/* TODO: Check this in the resulting binary */
+static void (*const compartment_destructor_ptr)(void) =
+    COMPARTMENT_IDENT(__wrap_ia2_compartment_destructor);
+
+void COMPARTMENT_IDENT(ia2_setup_destructors)(void) {
+  int res = 0;
+  Elf64_Ehdr *ehdr = &__ehdr_start;
+  Elf64_Phdr *phdr = (Elf64_Phdr *)((uint8_t *)ehdr + __ehdr_start.e_phoff);
+  assert(sizeof(Elf64_Phdr) == ehdr->e_phentsize);
+  Elf64_Phdr *dynamic_phdr = NULL;
+  for (int i = 0; i < ehdr->e_phnum; i++) {
+    if (phdr[i].p_type == PT_DYNAMIC) {
+      dynamic_phdr = &phdr[i];
+      break;
+    }
+  }
+  if (!dynamic_phdr) {
+    return;
+  }
+  Elf64_Dyn *dynamic =
+      (Elf64_Dyn *)((uintptr_t)&__ehdr_start + dynamic_phdr->p_vaddr);
+  void *rewrite_base = NULL;
+  void *finalizers_region =
+      (void *)IA2_ROUND_DOWN((uintptr_t)&finalizers, PAGE_SIZE);
+  size_t finalizers_region_size =
+      IA2_ROUND_DOWN(((uintptr_t)&finalizers) + sizeof(finalizers) + PAGE_SIZE,
+                     PAGE_SIZE) -
+      (uintptr_t)finalizers_region;
+  res = mprotect(finalizers_region, finalizers_region_size,
+                 PROT_READ | PROT_WRITE);
+  if (res != 0) {
+    fprintf(stderr, "Failed to set relro section writable: %d\n", res);
+    exit(1);
+  }
+  for (int i = 0; i < dynamic_phdr->p_memsz / sizeof(Elf64_Dyn); i++) {
+    if (dynamic[i].d_tag == DT_FINI || dynamic[i].d_tag == DT_FINI_ARRAY ||
+        dynamic[i].d_tag == DT_FINI_ARRAYSZ) {
+      uintptr_t dynamic_entry_addr = (uintptr_t)&dynamic[i].d_un.d_val;
+      bool in_finalizers_mapping =
+          dynamic_entry_addr >= (uintptr_t)finalizers_region &&
+          dynamic_entry_addr + sizeof(Elf64_Xword) <=
```

Comment:
This should be `<` since finalizers_region + region_size is on the next page

Reply:
I don't think so, `dynamic_entry_addr + sizeof(Elf64_Xword)` is the address of the next byte after the entry.

Reply:
nit: Might be more obvious that it's the next byte if we use `&dynamic[i + 1]` like in the assert below.